File: solutions/typescript/dnd-character/2/dnd-character.ts
export class DnDCharacter {
  public readonly strength = DnDCharacter.generateAbilityScore()
  public readonly dexterity = DnDCharacter.generateAbilityScore()
  public readonly constitution = DnDCharacter.generateAbilityScore()
  public readonly intelligence = DnDCharacter.generateAbilityScore()
  public readonly wisdom = DnDCharacter.generateAbilityScore()
  public readonly charisma = DnDCharacter.generateAbilityScore()
  public readonly hitpoints = DnDCharacter.getModifierFor(this.constitution) + 10

  public static generateAbilityScore(): number {
    let mini = DnDCharacter.diceRoll()
    let dice: number
    let total = 0
    for (let i = 0; i < 3; ++i) {
      dice = DnDCharacter.diceRoll()
      if (dice < mini) {
        total += mini
        mini = dice
      } else {
        total += dice
      }
    }
    return total
  }

  private static diceRoll(): number {
    return Math.random() * 6 + 1
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2)
  }
}

File: solutions/kotlin/dnd-character/2/src/main/kotlin/DndCharacter.kt
import kotlin.random.Random

class DndCharacter {

    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = modifier(constitution) + 10

    companion object {

        fun ability(): Int {
            var mini = Random.nextInt(1, 7)
            var total = 0
            var dice: Int
            for (i in 1..3) {
                dice = Random.nextInt(1, 7)
                if (dice < mini) {
                    total += mini
                    mini = dice
                } else {
                    total += dice
                }
            }
            return total
        }

        fun modifier(score: Int): Int {
            return Math.floorDiv(score - 10, 2) 
        }
    }

}
